R/usage_to_annual_nets_distributed.R

```diff
@@ -0,0 +1,234 @@
+    # for access levels beyond those observed
+    loess_for_prediction <- rbind(loess_for_prediction,
+                                  data.frame(iso3=NA, month=NA, access=1, 
+                                             percapita_nets = 1, loess_predicted_access=1))
+    print("Access-nets per capita curve is extrapolated linearly beyond observed levels")
+  } else {
+    print("No extrapolation beyond observed access-nets per capita relationship - return NA")
+  }
+  
+  max_usage <- country_use_rate[country_use_rate$iso3 %in% country_iso3,c("iso3","use_rate")]
+  names(max_usage)[names(max_usage)=="use_rate"] <- "maximum_usage"
+  
+  if (any(target_usage<=max(country_use_rate[country_use_rate$iso3 %in% country_iso3,"use_rate"]))==FALSE) {
+    print(max_usage)
+    stop("Target usage(s) cannot be achieved with estimated use rate in any of the countries. Refer to maximum usage above.")
+  } else if(any(target_usage>min(country_use_rate[country_use_rate$iso3 %in% country_iso3,"use_rate"]))) {
+    print(max_usage)
+    print("Target usage(s) cannot be achieved with estimated use rate in some of the countries - return NA. Refer to maximum usage above.")
+  }
+  
+  targets_all <- lapply(target_usage, function(this_use) {
+    targets_dt <- country_use_rate[country_use_rate$iso3 %in% country_iso3, 
+                                   c("iso3", "year", "use_rate")]
+    targets_dt$target_use <- this_use
+    # Convert target usage to target access: access = usage/use rate
+    # since use rate = the proportion of people with access to a net who slept under it
+    targets_dt$target_access <- targets_dt$target_use/targets_dt$use_rate
+    targets_dt <- targets_dt[order(targets_dt$target_access),]
+    targets_dt$target_percapita_nets <- NA
+    targets_dt$target_percapita_nets[
+      targets_dt$target_access<=max(loess_for_prediction$loess_predicted_access)] <- 
+      approx(loess_for_prediction$loess_predicted_access,
+             loess_for_prediction$percapita_nets, 
+             targets_dt$target_access[targets_dt$target_access<=max(loess_for_prediction$loess_predicted_access)])$y
+    # Relationship between access and NPC is aggregated for all countries,
+    # but use rate to convert usage to access is country-specific (derived from spatiotemporal regression?)
+    return(targets_dt[order(targets_dt$iso3),])
+  })
+  targets_all <- do.call("rbind", targets_all)
+  targets_all$target_access[targets_all$target_access>1] <- NA
+  # Set access to NA where it is over 1 and excluded from conversion
+  
+  ### 2. Convert equilibrium nets per capita to annual nets distributed per capita ###
+  
+  # Read in net half lives (median retention times) for each country
+  half_life_data <- data$half_life_data
+  
+  # Estimate time at which all nets are lost (parameter l) based on net half life in each country
+  half_life_data$l <- half_life_data$half_life/sqrt(1 - k / (k - log(0.5)))
```

Comment:
This is quite a large function and a number of sections can be split out into their own small functions that perform single, simple tasks. These are then much easier to unit test.